lfs/diff_index_scanner.go

```diff
@@ -0,0 +1,194 @@
+}
+
+// NewDiffIndexScanner initializes a new `DiffIndexScanner` scanning at the
+// given ref, "ref".
+//
+// If "cache" is given, the DiffIndexScanner will scan for differences between
+// the given ref and the index. If "cache" is _not_ given, DiffIndexScanner will
+// scan for differences between the given ref and the currently checked out
+// tree.
+//
+// If any error was encountered in starting the command or closing its `stdin`,
+// that error will be returned immediately. Otherwise, a `*DiffIndexScanner`
+// will be returned with a `nil` error.
+func NewDiffIndexScanner(ref string, cached bool) (*DiffIndexScanner, error) {
+	cmd, err := startCommand("git", diffIndexCmdArgs(ref, cached)...)
+	if err != nil {
+		return nil, errors.Wrap(err, "diff-index")
+	}
+
+	if err = cmd.Stdin.Close(); err != nil {
+		return nil, errors.Wrap(err, "diff-index: close")
+	}
+
+	return &DiffIndexScanner{
+		from: bufio.NewScanner(cmd.Stdout),
+	}, nil
+}
+
+// diffIndexCmdArgs returns a string slice containing the arguments necessary
+// to run the diff-index command.
+func diffIndexCmdArgs(ref string, cached bool) []string {
+	args := []string{"diff-index", "-M"}
+	if cached {
+		args = append(args, "--cached")
+	}
+	args = append(args, ref)
+
+	return args
+}
+
+// Scan advances the scan line and yields either a new value for Entry(), or an
+// Err(). It returns true or false, whether or not it can continue scanning for
+// more entries.
+func (s *DiffIndexScanner) Scan() bool {
+	if !s.prepareScan() {
+		return false
+	}
+
+	s.next, s.err = s.scan(s.from.Text())
+	if s.err != nil {
+		s.err = errors.Wrap(s.err, "diff-index scan")
+	}
+
+	return s.err == nil
+}
+
+// Entry returns the last entry that was Scan()'d by the DiffIndexScanner.
+func (s *DiffIndexScanner) Entry() *DiffIndexEntry { return s.next }
+
+// Entry returns the last error that was encountered by the DiffIndexScanner.
+func (s *DiffIndexScanner) Err() error { return s.err }
+
+// prepareScan clears out the results from the last Scan() loop, and advances
+// the internal scanner to fetch a new line of Text().
+func (s *DiffIndexScanner) prepareScan() bool {
+	s.next, s.err = nil, nil
+	if !s.from.Scan() {
+		s.err = s.from.Err()
+		return false
+	}
+
+	return true
+}
+
+// scan parses the given line and returns a `*DiffIndexEntry` or an error,
+// depending on whether or not the parse was successful.
+func (s *DiffIndexScanner) scan(line string) (*DiffIndexEntry, error) {
+	parts := strings.Split(line, "\t")
+	if len(parts) < 2 {
+		return nil, errors.Errorf("invalid line: %s", line)
+	}
+
+	desc := strings.Fields(parts[0])
+	if len(desc) < 5 {
+		return nil, errors.Errorf("invalid description: %s", parts[0])
+	}
+
+	entry := &DiffIndexEntry{
+		SrcMode: strings.TrimPrefix(desc[0], ":"),
+		DstMode: desc[1],
+		SrcSha:  desc[2],
+		DstSha:  desc[3],
+		Status:  DiffIndexStatus(rune(desc[4][0])),
+		SrcName: parts[1],
+	}
+
+	if score, err := strconv.Atoi(desc[4][1:]); err != nil {
```

Comment:
It's subtle, but using `strings.Fields` does guarantee that at each field is at least one, non-whitespace character long.

Since we check that the `len(desc)` is at least 5, we're guaranteed to have at least 5 non-empty slice elements.

[Here](https://play.golang.org/p/uvyIM5YDnt) is an example.